### DISCUSSION TITLE
[Declined] Add synchronous alternatives to API

### DISCUSSION THREAD

**Discussion by lokka30** [ArcanePlugins/Treasury#185] [Ideas]
Good day everyone! I've arrived to start yet another controversial conversation, though I strongly believe that this is an important factor for the future of the library hereon.

---

I am almost certain that the current lack of simple synchronous methods in Treasury's economy API is severely throttling the potential adoption of the library.

Ideally, we have async and sync methods in the API so developers can choose either.

I believe that their choice is in majority based upon these factors:

1. whether they are capable of writing concurrent code;
2. whether they are willing to invest time in learning to write concurrent code;
3. whether they are willing to refactor their existing non-concurrent code.

I completely agree that developers very well *should* be using Treasury in a concurrent manner. Although, the reality is that most developers do not fit all these criteria – some are willing to overcome it (e.g., how some of the current Treasury adopters have), but I believe that most are not. This leads into two further factors to consider:

1. The only significant selling point that Treasury has for the majority of plugins, is the thorough support for concurrency. Unfortunately, it's evident that a low proportion of developers are interested in writing concurrent programs.

2. A plethora of developers on the platform are young and inexperienced. There should be only little expectation that these developers are able to learn how to write concurrent code, which is somewhat complicated to do.

**Comment by lokka30** [comment]
Interested to hear thoughts of anyone reading this. 😃 
Pinging the regulars (let me know to opt-out): @MrIvanPlays @Jikoo @MrNemo64 @yannicklamprecht

**Comment by MrNemo64** [comment]
I thought we were kinda like forcing good code. They can always use the CompletableFuture#get to get the value, although I don't expect newies to know how to unwrap a Future. 

I don't have strong feelings towards this idea anyways.

**Reply by lokka30** [reply]
Good point about `get` - I guess developers can just use it if they want to go with a non-concurrent approach and deal with worse looking code as a side effect.

I did forget about our mission to promote better code when writing this post, it's something to also consider if anything changes based upon this discussion.

Thanks for the reply. 😀

**Reply by MrNemo64** [reply]
> Good point about `get` - I guess developers can just use it if they want to go with a non-concurrent approach and deal with worse looking code as a side effect.
> 
> I did forget about our mission to promote better code when writing this post, it's something to also consider if anything changes based upon this discussion.
> 
> Thanks for the reply. 😀

To be honest, is we want more users we should develop the other APIs. It's been long enough since we released the economy one and we have a long list.

**Reply by lokka30** [reply]
> > Good point about `get` - I guess developers can just use it if they want to go with a non-concurrent approach and deal with worse looking code as a side effect.
> > I did forget about our mission to promote better code when writing this post, it's something to also consider if anything changes based upon this discussion.
> > Thanks for the reply. 😀
> 
> To be honest, is we want more users we should develop the other APIs. It's been long enough since we released the economy one and we have a long list.

I'm thinking of this order:

`Develop & release v2` ➡️ `Write documentation` ➡️ `Start work on permissions API`

I'll have my freedom in a few days time to get back to working on the code.

**Reply by yannicklamprecht** [reply]
The only thing I would add as a module, is a syncronise wrapper, that basically calls #get().
So that you do not integrated that into the main API.

**Reply by lokka30** [reply]
> The only thing I would add as a module, is a synchronous wrapper, that basically calls `#get()`, so that you do not integrate that into the main API.

Sounds like a brilliant way to satisfy both parties.

**Reply by MrNemo64** [reply]
> > > Good point about `get` - I guess developers can just use it if they want to go with a non-concurrent approach and deal with worse looking code as a side effect.
> > > I did forget about our mission to promote better code when writing this post, it's something to also consider if anything changes based upon this discussion.
> > > Thanks for the reply. 😀
> > 
> > 
> > To be honest, is we want more users we should develop the other APIs. It's been long enough since we released the economy one and we have a long list.
> 
> I'm thinking of this order:
> 
> `Develop & release v2` ➡️ `Write documentation` ➡️ `Start work on permissions API`
> 
> I'll have my freedom in a few days time to get back to working on the code.

Alright. Is there any developed idea with the permission and chat API?

**Reply by MrIvanPlays** [reply]
why does v2 get released before a new api? we need permissions api in v2 not just remove deprecations. also the documentation should also be paired to v2. of course you can not do it like that but then why would you want an undocumented, kinda featureless major version bump?

**Reply by MrNemo64** [reply]
> why does v2 get released before a new api? we need permissions api in v2 not just remove deprecations. also the documentation should also be paired to v2. of course you can not do it like that but then why would you want an undocumented, kinda featureless major version bump?

Agree with this

**Comment by Jikoo** [comment]
Personally, the reason I have yet to adopt Treasury is not due to difficulty, it's because I have yet to write a generified economy hook system, which was my [original request](https://www.spigotmc.org/threads/released-treasury-economy-api.527805/#post-4276314) when hearing about it. I recently wrote a [system for handling optional dependencies](https://github.com/Jikoo/PlanarWrappers/blob/44f0aecd3e817e8d9267b585621a83b1f12bf04c/src/main/java/com/github/jikoo/planarwrappers/service/ProvidedService.java) and added a [Vault handler](https://github.com/Jikoo/PlanarWrappers/blob/44f0aecd3e817e8d9267b585621a83b1f12bf04c/src/main/java/com/github/jikoo/planarwrappers/service/VaultEconomy.java) a little while ago. Treasury isn't much more difficult and combining the two has a few fiddly design choices, but it's now on hold for 1.19 and whatnot. When I get that done my adoption will be a lot simpler.

**Reply by lokka30** [reply]
I feel like an automatic Treasury/Vault handler is something we could undertake to (relatively) rapidly increase adoption of the APIs (even indirectly).

My thought is to package that as a small shaded utility, separate from the Treasury project.

**Comment by MrIvanPlays** [comment]
Strongly disagree on adding sync methods. Do you even have the slightest clue of what this can cause?

**Reply by lokka30** [reply]
Nope.

**Reply by MrIvanPlays** [reply]
here's a (not full) list of what can happen when doing a database call synchronously on Bukkit's main thread:

1. Server crash
2. Connection drop
3. Watchdog trigger - this is not _as bad as_ a server crash, but it is close, and even though it is _not_ a crash, it 99% leads to server quit, which is basically the server shutting down in _kinda_ a normal way, but not exactly
4. TPS loss - as you may know, synchronous calls onto a main thread will cause the main thread to block until the synchronous call finishes, which directly means TPS loss. Now this can go real bad real fast. See, each time the server lost tps, it needs to keep up, and if it can't keep up you have some big problems with lag
5. Depending on the period of database calls, can cause lag stutters due to aforementioned problems

these are kinda the main reasons why I/O operations and databases are done asynchronously. it is better to wait for the outcome rather to block a thread. and that's what basically futures do. if a developer doesn't like it or doesn't know it and shows no interest of learning concurrency, it's their problem and we shouldn't care for such bad developers.

**Reply by lokka30** [reply]
Thank you very much. Hopefully this will help persuade developers to get off their ass and fix their code :)

**Reply by lokka30** [reply]
I agree with Ivan here. @MrNemo64 and @Jikoo what about you?

**Reply by MrNemo64** [reply]
If a developer really doesn't want to learn how to write async code, they'll just (ab)use the `#get()` method and will just cause all these problems. At least right now we don't make it easyer to fall into these issues

**Reply by Jikoo** [reply]
I/O on the main server thread is undesirable. If you firmly believe it will significantly increase adoption I'm not going to hold you back, but I don't think it's a good idea.

If you want to invest efforts into encouraging wider adoption, imo the best use of your time would be to PR Treasury support into major economy providers and consumers. Proper implementations and usages can then (hopefully) serve as an example for less-experienced developers.

**Reply by MrIvanPlays** [reply]
we're not adding the CompletableFuture by default then if you are having concerns of the (ab)use of `#get` and going other routes:
1. No change to current concurrency whatsoever OR
2. Make a simplifiedfuture, what i suggested back when #37 was around

**Reply by MrIvanPlays** [reply]
> I/O on the main server thread is undesirable. If you firmly believe it will significantly increase adoption I'm not going to hold you back, but I don't think it's a good idea.
> 
> If you want to invest efforts into encouraging wider adoption, imo the best use of your time would be to PR Treasury support into major economy providers and consumers. Proper implementations and usages can then (hopefully) serve as an example for less-experienced developers.

Completely agree with this but what about what i posted literally 30 sec ago?

**Reply by Jikoo** [reply]
If we think a simplified future would help with adoption, sure. Definitely needs careful thought during design though - both of us missed [unsafe usages](https://github.com/lokka30/Treasury/pull/180#discussion_r864841786) due to the fact that default futures do not mandate error handling.

**Reply by MrIvanPlays** [reply]
its not gonna feature a `#get` and a `#join` method - easy.

**Comment by MrIvanPlays** [comment]
Wait why are we releasing 2.0.0 before a new api?? It is a major version
increase after all, we need more than just removing deprecations.

На пт, 10.06.2022 г., 18:35 Lachlan ***@***.***> написа:

**Reply by lokka30** [reply]
> Wait why are we releasing 2.0.0 before a new api?? It is a major version increase after all, we need more than just removing deprecations. На пт, 10.06.2022 г., 18:35 Lachlan ***@***.***> написа:
> […](#)
> Good point about get - I guess developers can just use it if they want to go with a non-concurrent approach and deal with worse looking code as a side effect. I did forget about our mission to promote better code when writing this post, it's something to also consider if anything changes based upon this discussion. Thanks for the reply. 😀 To be honest, is we want more users we should develop the other APIs. It's been long enough since we released the economy one and we have a long list. I'm thinking of this order: Release v2 ➡️ Write documentation ➡️ Start work on permissions API I'll have my freedom in a few days time to get back to working on the code. — Reply to this email directly, view it on GitHub <[#185 (reply in thread)](https://github.com/lokka30/Treasury/discussions/185#discussioncomment-2924143)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJBTDWJL27AUJOSMG6R6AK3VONODZANCNFSM5YNGWIAQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I didn't plan on releasing v2 before new APIs, nor prior to new documentation being written.
